### PR TITLE
[GraphTrainer] Make model cudagraphable by eliminating `lift_fresh_copy` nodes in TP>1

### PR DESCRIPTION
**Pull request by WhoisZihan**
# Motivation
When training with GraphTrainer, models with CrossEntropyLoss(e.g., llama3 and qwen3) will produce non-cudagraphable traced fx graph with TP>1. The reason is that some nodes in loss calculation involve cpu operation, preventing graph capture.

> Setup: 2 x A100, pytorch nightly 0629
> Model: llama3 debugmodel, qwen3 debugmodel
> Dataset: c4_test
> Config: FSDP=1, TP=2

# Identify the cpu nodes
By traversing the graph, I discover these`lift_fresh_copy` nodes generated by loss index_put logic.

Current label filtering logic such as [local_result[out_of_range.unsqueeze(-1)] = 0](https://github.com/pytorch/torchtitan/blob/0f5dfc1e90c42115313e5e39a49537c4edef3279/torchtitan/components/loss.py#L173) will invoke interfaces like `index_put_`, which [**only accepts values of Tensor type**](https://docs.pytorch.org/docs/2.12/generated/torch.Tensor.index_put_.html). Therefore, it will generate `lift_fresh_copy` node in fx graph that constructs tensor from scalar, which is not cudagraph-able.

One option is to replace the `index_put` with `masked_fill` which accepts python scalar as values. However, since lift_fresh_copy with constant does not have any side effects, we can safely capture these nodes into cudagraph.

# Benchmark

llama3 debugmodel
| config | before tps |  after tps | speedup |
|------|--------|------|-------|
| TP=2 Flex | 113578 | 206566 | **1.82x** |
| TP=2 SDPA | 130333 | 207143 | **1.55x** |

qwen3 debugmodel
| config | before tps |  after tps | speedup |
|------|--------|------|-------|
| TP=2 Flex | 74773 | 78769 | **1.05x** |

## For those with stable pytorch (<= 2.12.1 at the moment)
If you want to make the training fully cudagraphable on current stable pytorch, you need to additionally pick https://github.com/pytorch/torchtitan/pull/3818/changes/b8bdf9a380ea05040cb206ac99188995b8574319 to eliminate the dead cpu nodes produced by improper tracing of DShard propagation, which is fixed in nightly pytorch https://github.com/pytorch/pytorch/pull/185865 as suggested by @aditvenk
 
<details>
  <summary>Detailed dead cpu nodes</summary>

## 1. lm_head weight in ChunkedCELoss, the first chunk

The sequence of llama3 debugmodel in splitted into 8 chunk in ChunkedCELoss, and each chunk goes through lm_head projection. I find that the first chunk backward phase generates dead chain `(empty_strided + empty_strided_1) -> add`, where the add node is not consumed by any other node.

```
        # Annotation: {'module_fqn': 'loss'} File: /home/hadoop-quanta/venv/lib/python3.12/site-packages/torch/distributed/tensor/_redistribute.py:1939 in forward, code: output, spec = _redistribute_backward(
        _to_copy_84: "f32[1024, 256][256, 1]cuda:0" = torch.ops.aten._to_copy.default(mm_40, dtype = torch.float32);  mm_40 = None

        # Annotation: {'module_fqn': 'loss'} No stacktrace found for following nodes
        empty_strided_58: "f32[2048, 256][256, 1]cpu" = torch.ops.aten.empty_strided.default([2048, 256], [256, 1], dtype = torch.float32, device = device(type='cpu'), pin_memory = False)
        empty_strided_59: "f32[2048, 256][256, 1]cpu" = torch.ops.aten.empty_strided.default([2048, 256], [256, 1], dtype = torch.float32, device = device(type='cpu'), pin_memory = False)
        add__1: "f32[2048, 256][256, 1]cpu" = torch.ops.aten.add_.Tensor(empty_strided_58, empty_strided_59);  empty_strided_58 = empty_strided_59 = add__1 = None
        add__2: "f32[1024, 256][256, 1]cuda:0" = torch.ops.aten.add_.Tensor(_to_copy_78, _to_copy_84);  _to_copy_78 = _to_copy_84 = None
```

Total fx nodes and seq_nr for all chunks are below
NOTE. the debugmodel happens to have same size of vocab and seqlen(both 2048), but I verified that the shape of dead node empty_strided is `[vocab, hidden_dim]` by changing the vocab size.

| node | seq_nr | device | shape | users | comments |
|------|--------|------|-------|-------|------|
| `add__1` **(DEAD)** | **787** | cpu | `[2048,256]`= full vocab | 0 | dead phantom |
| `add__2` (chunk0+chunk1) | **787** | cuda | `[1024,256]` | 1 | read `p.grad.add_(g)` |
| `add__3` (chunk1+chunk2) | 806 | cuda | `[1024,256]` | 1 | |
| `add__4` | 825 | cuda | `[1024,256]` | 1 | |
| `add__5` | 844 | cuda | `[1024,256]` | 1 | |
| `add__6` | 863 | cuda | `[1024,256]` | 1 | |
| `add__7` | 882 | cuda | `[1024,256]` | 1 | |
| `add__8` | 901 | cuda | `[1024,256]` | 1 | |

## 2. layer.0.attention_norm - first norm after tok_embedding
When running with TP=2, input of `tok_embedding` is FakeTensor but the output becomes DTensor, with the calling stack

```
tok_embeddings: local_map(F.embedding) → plain tensor
                → output redistribute (Partial → Shard(1))
                → First SP DTensor
                ↓
layers.0.attention_norm(x)
```

Each layer comes with an attention norm in the beginning. The first `F.rms_norm` under make_fx and FakeTensor without local_map produces the dead cpu node chain.
```
empty_strided[8,2048,256]cpu -> _to_copy_4 -> pow_1(x^2) -> mean -> add_(+eps) # add_ is dead, 0 consumer
``` 

The later layers' inputs comes from previous layers(attention/ffn with redistribution), thus, rms_norm consistently dispatches to actual `_fused_rms_norm` with correct cuda device.
```
attention (QKV internal Replicate → matmul → wo RowwiseParallel reduce-scatter)
→ residual add
→ ffn → residual add
→ next attention_norm
```

## Benchmark

NOTE. this bench result only apply to stable pytorch(till 2.12.1 so far)
I tested 3 sets of configurations, each runs 30 steps and measure the average of 10-30 steps.
- A: current implementation, with dead cpu nodes
- B: elimination of dead cpu nodes, but no cudagraph
- C: elimination of dead cpu ndoes, with cudagraph

llama3 debugmodel
| config | A TPS |  B TPS | C TPS | A->B Improv. | A->C Improv. |
|------|--------|------|-------|-------|-------|
| FSDP=2/TP=1 Flex | 197423 | 252570 | 362027 | **1.28x** | **1.83x** |
| FSDP=1/TP=2 Flex | 55884 | 113528 | 206502 | **2.03x** | **3.70x** |
| FSDP=1/TP=2 SDPA | 59002 | 132074 | 204828 | **2.24x** | **3.47x** |

qwen3 debugmodel
| config | A TPS |  B TPS | C TPS | A->B Improv. | A->C Improv. |
|------|--------|------|-------|-------|-------|
| FSDP=2/TP=1 | 98413 | 98991 | 100600 | **1.01x** | **1.02×** |
| FSDP=1/TP=2 | 17497 | 74602 | 78103 | **4.26×** | **4.46×** |

## Question: why does `eliminate_dead_node_pass` not eliminate these *dead* nodes?
Current graph.eliminate_dead_code will use [node.is_pure](https://github.com/pytorch/pytorch/blob/main/torch/fx/node.py#L760) to decide if we dan safely erase a node. But it only checks placeholder, output and impure module/function, not covering such dead cpu node chains. Because these nodes are not placeholder or module.
</details>

### CODE REVIEW

**Comment by aditvenk** [comment]
@WhoisZihan  -- could you please re-try with latest PyTorch nightly?

We merged a PT change to not trace into DTensor shard propagation and could affect your observations here. 
https://github.com/pytorch/pytorch/pull/185865

**Review by aditvenk — CHANGES_REQUESTED** [review]
Pl re-try with latest PT nightly

**Comment by WhoisZihan** [comment]
> @WhoisZihan -- could you please re-try with latest PyTorch nightly?
> 
> We merged a PT change to not trace into DTensor shard propagation and could affect your observations here. [pytorch/pytorch#185865](https://github.com/pytorch/pytorch/pull/185865)

@aditvenk Just reran with nightly. The empty_strided dead nodes are gone both for FSDP and TP, but `lift_fresh_copy` nodes still exist under TP due to the `_cross_entropy_via_local_map` path calling `_LossParallelCrossEntropy`.

Do you want me to revert part of the dead node detection and only keep the loss masked_fill part, or start a new one?

**Comment by WhoisZihan** [comment]
**performance statistics with pytorch nightly**

With pytorch nightly, FSDP is full cudagraphable now, so only TP results included. The performance comparison against the loss fix [501f65b](https://github.com/pytorch/torchtitan/pull/3818/commits/501f65bf7fc745bacb2fd23b0a26aeb569409cfb).

llama3 debugmodel
| config | before tps |  after tps | speedup |
|------|--------|------|-------|
| TP=2 Flex | 113578 | 206566 | **1.82x** |
| TP=2 SDPA | 130333 | 207143 | **1.55x** |

qwen3 debugmodel
| config | before tps |  after tps | speedup |
|------|--------|------|-------|
| TP=2 Flex | 74773 | 78769 | **1.05x** |

This is pretty much aligned with the B->C speedup in the original test.

**Comment by aditvenk** [comment]
> > @WhoisZihan -- could you please re-try with latest PyTorch nightly?
> > We merged a PT change to not trace into DTensor shard propagation and could affect your observations here. [pytorch/pytorch#185865](https://github.com/pytorch/pytorch/pull/185865)
> 
> @aditvenk Just reran with nightly. The empty_strided dead nodes are gone both for FSDP and TP, but `lift_fresh_copy` nodes still exist under TP due to the `_cross_entropy_via_local_map` path calling `_LossParallelCrossEntropy`.
> 
> Do you want me to revert part of the dead node detection and only keep the loss masked_fill part, or start a new one?

Hi, please revert the unnecessary changes in this PR and you can update this one (including the PR description). Thanks!

**Comment by WhoisZihan** [comment]
@aditvenk Hi Aditya, I just reverted the dead node detection logic, also refined the PR title and description. I decide to preserve the section of dead cpu nodes in case someone wants(or needs) to use stable pytorch, but I make it collapsible in the description.

Thanks for reviewing my pr!

File: torchtitan/components/loss.py [hunk 1]
```diff
@@ -162,22 +162,26 @@ def forward(
...
-        local_labels[out_of_range] = 0
+        local_labels.masked_fill_(out_of_range, 0)
 
         local_result = torch.gather(log_probs, -1, local_labels.unsqueeze(-1))
-        local_result[out_of_range.unsqueeze(-1)] = 0
-        local_result = funcol.all_reduce(
-            local_result, reduceOp=dist.ReduceOp.SUM.name, group=tp_group
-        )
+        local_result.masked_fill_(out_of_range.unsqueeze(-1), 0)
+        dist.all_reduce(local_result, op=dist.ReduceOp.SUM, group=tp_group)
```
Comment (sanketpurandare) [inline torchtitan/components/loss.py]: We would want to use functional collectives

Comment (WhoisZihan) [reply torchtitan/components/loss.py]: Recovered, thanks. Is the functional collective preferred over legacy ones for better traceability? The legacy dist interface are still extensively used across a lot of projects. If they are legacy, does it mean we will have to switch to functional collectives someday in the future?

File: torchtitan/components/loss.py [hunk 2]
```diff
@@ -162,22 +162,28 @@ def forward(
 
         # Mask labels outside this vocab shard; the TP all-reduce below selects
         # the owner rank's log probability for each target token.
-        safe_labels = torch.where(labels_1d != IGNORE_INDEX, labels_1d, 0)
+        # NOTE: use masked_fill/masked_fill_ (Scalar overload) rather than
```
Comment (aditvenk) [inline torchtitan/components/loss.py]: I don't think changing loss in core is merited here for what is essentially a make_fx / graph_trainer limitation. The current code is cudagraphable in eager, nothing wrong with it. 

What is breaking CG in graph_trainer is is_cudagraphable() rejecting the fresh_lift_copy node for the scalar 0. But I don't see why this should reject -- nothing is "unsafe" about this scenario, a scalar 0 is fed to an op `index_put_` -- it is not mutated or anything. 

I can imagine two approaches to make this cudagraphable, local to graph trainer:
1. Allow fresh_lift_copy as is_cudagraphable conservatively when it is safe to do
2. Rewrite the fresh_lift_copy -> index_put sequence into a masked_fill as a graph pass.

**Review by aditvenk — CHANGES_REQUESTED** [review]
Let's avoid changing loss in core. Suggested some approaches to restrict change to GraphTrainer alone. 
Also, pl add a test for the change if you choose to implement differently.

File: torchtitan/components/loss.py (hunk 2 above)
Comment (WhoisZihan) [reply torchtitan/components/loss.py]: Makes sense, I've modified the is_cudagraphable to support lift_fresh_copy with constant input. Such constant input nodes have name `_tensor_constantN` and op type `get_attr` because it is baked as an attribute of the graph. In comparison, normal cpu tensor will become `placeholder`.

**Comment by WhoisZihan** [comment]
> Let's avoid changing loss in core. Suggested some approaches to restrict change to GraphTrainer alone. Also, pl add a test for the change if you choose to implement differently.

Thanks for the suggestion. I've enabled the capture of lift_fresh_copy in the `is_cudagraphable` check if possible. Do you have further comment on current implementation?

**Comment by WhoisZihan** [comment]
@aditvenk Hi, would you help review the latest fix? The implementation based on your suggestion is small and straightforward. Thanks.
